### PR TITLE
feat(review): add finding quality gates

### LINKED ISSUE #41: review: Claude·Gemini finding 단위 근거 검증과 soft filtering
## 배경

automation v1.45.2가 적용된 pim-check PR #101에서 Gemini가 현재 지원 경로에 존재하지 않는 plan_global 동작을 HIGH로 지적하고, 물질적 영향이 입증되지 않은 소형 YAML 중복 로딩을 MEDIUM으로 지적했다. 후속 delta 재리뷰에서는 철회됐지만 최초 리뷰 게시 단계에서는 prompt-only 규칙을 통과했다.

- 사례: https://github.com/jhw7500/pim-check/pull/101
- 처분 근거: https://github.com/jhw7500/pim-check/pull/101#issuecomment-5385042772

## 목표

Claude·Gemini 출력을 finding 단위로 파싱하고, 검증 가능한 근거가 부족한 candidate만 actionable 리뷰에서 제외한다. 개별 품질 오류는 리뷰 전체 실패로 만들지 않고, 입력·상태·scope 무결성 오류만 기존처럼 fail-closed 처리한다.

## 정책

- Changed anchor는 sealed PR scope의 실제 추가 줄이어야 한다.
- trigger evidence는 현재 저장소의 실제 경로·줄·인용문과 일치해야 한다.
- HIGH/MEDIUM은 concrete trigger, impact class, material impact를 요구한다.
- 조건을 만족하지 못한 finding은 제외하고 필터 수와 모델 주장 최고 심각도를 투명하게 기록한다.
- 추가 모델/API 호출은 하지 않는다.
- diff/head/hash 불일치, 출력 전체 파싱 불가, 잘림·빈 응답은 라운드 전체 실패를 유지한다.

## 완료 조건

- 유효한 finding은 보존된다.
- 근거 없는 개별 finding은 게시되지 않는다.
- 모든 candidate가 제외돼도 No validated blocking issues found로 표현하며 필터 수를 함께 표시한다.
- 이전 정상 checkpoint와 current-head 무결성 계약을 훼손하지 않는다.
- Claude·Gemini 양쪽에 동일한 정책과 회귀 테스트가 적용된다.

### LINKED ISSUE #42: review: 재리뷰 상태 분류를 이전 finding identity에 결속
## 배경

pim-check PR #101의 Claude 최종 재리뷰는 이전 라운드에 actionable finding이 없었는데도, 이전에 긍정적으로 평가했던 변경을 Resolved로 분류했다. 현재 Claude·Gemini는 Still open, Resolved, Retracted를 이전 정상 리뷰의 finding과 기계적으로 연결하지 않는다.

- 사례: https://github.com/jhw7500/pim-check/pull/101#issuecomment-5385024910

## 목표

재리뷰 carryover 상태가 인증된 이전 finding identity에만 연결되도록 정규화한다.

## 정책

- 모든 actionable finding에 라운드 내 고유 ID를 부여한다.
- Still open, Resolved, Retracted는 이전 성공 checkpoint의 ID와 정확히 연결돼야 한다.
- Resolved는 code delta가 있는 경우에만 허용한다.
- Retracted는 이전 주장이 현재 증거로 반증된 경우에만 허용한다.
- 이전 ID가 없는 상태 항목은 actionable 상태에서 제거하고 필요하면 일반 요약으로 이동한다.
- Gemini prompt의 Cannot verify 생성 지시와 최종 출력 금지 규칙의 모순을 제거한다.
- 개별 상태 오분류는 soft normalization하고, 전체 문서 경계가 모호할 때만 라운드를 실패시킨다.

## 완료 조건

- 이전 finding 없는 가짜 Resolved가 게시되지 않는다.
- 올바른 Still open, Resolved, Retracted는 보존된다.
- 중복·모호한 ID는 안전하게 거부된다.
- 첫 라운드에 carryover 섹션을 생성할 수 없다.
- Claude·Gemini 동작이 공통 테스트로 고정된다.

### PR DESCRIPTION
**Pull request by jhw7500**
Closes #41
Closes #42
Relates to #43

## Summary

- Add a shared provider-independent review scope validator and canonicalizer.
- Soft-filter unsupported individual findings while preserving mechanically evidenced findings.
- Bind carryover to workflow-owned RVW IDs and authenticated prior canonical state.
- Migrate Claude and Gemini publication to schema 3 canonical-only comments.
- Preserve OpenCode schema 2 and prepare-only behavior.
- Add the v1.46 closed release inventory, historical v1.45.2 fixtures, and commit-only release verification.
- Add actionlint compatibility coverage for self-repository composite actions.

## Validation

- Full repository suite: 1293 passed, 48 subtests passed.
- Focused quality suite: 963 passed.
- YAML BaseLoader: 35 documents passed.
- actionlint 1.7.12: passed with no diagnostics.
- v1.46 commit-only verifier: passed at c6d0eea0e97181b407ebd7db7347404bf7d6076c.
- origin/main...HEAD whitespace and planned-path audit: passed.

Issue #43 remains open until v1.46 is released and both full and delta pim-check canaries are recorded.

### CODE REVIEW

**Comment by jhw7500** [comment]
@codex review

File: .github/actions/canonicalize-review/canonicalize_review.py [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+                        True, len(accepted) + len(still_open), filtered, normalized, maximum, "",
+                        tuple(sorted(reasons, key=lambda item: item.index)),
+                    )
+                    _write_atomic(
+                        request.canonical_file,
+                        _render_document(
+                            [_PriorFinding(
+                                stable_finding_id(request.reviewer, finding.anchor, finding.severity, finding.title),
+                                finding,
+                            ) for finding in accepted],
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/actions/canonicalize-review/canonicalize_review.py]: Fixed in ef124ed. Canonicalization now builds one `active_ids` set from authenticated prior active IDs, adds each newly generated stable ID as it is accepted, and normalizes any duplicate as `duplicate_prior_binding` before rendering. This covers both duplicate new blocks and a new block colliding with a carried active finding, so the canonical body cannot be poisoned for the next delta. The focused regression tests (including reloading the rendered body as prior state) pass: 2 passed.

**Comment by jhw7500** [comment]
@codex review

File: .github/actions/canonicalize-review/canonicalize_review.py [hunk 2]
```diff
@@ -0,0 +1,957 @@
...
+        if (raw_basis is None or raw_basis["kind"] not in {"measured", "unbounded-amplification"}
+                or not isinstance(raw_basis["path"], str) or not _safe_integer(raw_basis["line"])
+                or not isinstance(raw_basis["quote"], str)):
+            raise ScopeValidationError("previous performance basis is invalid")
+        basis = (raw_basis["kind"], TriggerEvidence(
+            raw_basis["path"], raw_basis["line"], raw_basis["quote"],
+        ))
+    elif bases:
+        raise ScopeValidationError("previous canonical finding is invalid")
+    prose = tuple(line for line in block.lines if line and not line.startswith("- "))
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/actions/canonicalize-review/canonicalize_review.py]: Fixed in 77d349a7a7b779af1353ce7bc4e2564ee9e86183. `_finding_prose()` now excludes only the closed structured-field prefixes and is shared by current and prior parsing, so supplemental bullets round-trip byte-for-byte. Added `test_bulleted_supplemental_prose_round_trips_as_authenticated_prior`; focused canonicalizer tests (118) and the full suite (2,382 tests, 48 subtests) pass.

**Comment by jhw7500** [comment]
@codex review

File: .github/workflows/claude-code-review.yml [hunk 3]
```diff
@@ -461,38 +569,85 @@ jobs:
...
+            const records = [];
+            for (const record of syntacticRecords) {
+              try {
+                const { data: run } = await github.rest.actions.getWorkflowRunAttempt({
+                  owner: context.repo.owner, repo: context.repo.repo,
+                  run_id: record.state.run_id, attempt_number: record.state.run_attempt,
+                });
+                if (runMatches(record, run)) records.push(record);
+              } catch (error) {
+                core.warning(`Ignoring Claude state without valid run provenance: ${error.message}`);
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/workflows/claude-code-review.yml]: Fixed in 64f96c4. Publication now fails closed when exact-attempt provenance is uncertain (no status, 403, 429, 500, or 503), while a definitive 404 skips only the absent record. Candidates are checked newest-first, lookup stops after the highest state authenticates, and duplicate generations prefer the larger comment ID. The Claude and Gemini paths share the same contract. Validation: 16 focused provenance tests, 85 upsert tests, 438 release-contract tests, actionlint 1.7.12, and the full suite (2398 passed, 48 subtests passed).

**Comment by jhw7500** [comment]
@codex review

File: .github/workflows/claude-code-review.yml [hunk 4]
```diff
@@ -190,23 +213,65 @@ jobs:
...
-          # only the validated state, while the model receives sanitized human-readable prose.
-          PREV_REVIEW="$(printf '%s\n' "$PREV_RECORD" | jq -r '.comment.body // ""')"
+          PREV_RECORD='null'
+          while IFS= read -r candidate; do
+            candidate_run_id="$(jq -r '.state.run_id' <<< "$candidate")"
+            candidate_attempt="$(jq -r '.state.run_attempt' <<< "$candidate")"
+            candidate_head="$(jq -r '.state.attempt_head' <<< "$candidate")"
+            run_record="$(gh api \
+              "repos/${GITHUB_REPOSITORY}/actions/runs/${candidate_run_id}/attempts/${candidate_attempt}" \
+              2>/dev/null || true)"
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/workflows/claude-code-review.yml]: Fixed in 06ed9331af28b886b52ff610fc619f396be5fa6b. Both Claude and Gemini collectors now fetch exact run-attempt provenance with response headers, treat only HTTP 404 as definitive absence, and abort on transport errors, 403/429/5xx, non-200 responses, or invalid payloads instead of falling back to older/no state. The issue-comment snapshot is also required, and sticky publication is gated on a successful collector outcome, so collection uncertainty cannot mutate prior state. Duplicate generations now prefer the larger comment ID. Verification: 18 focused fail-closed tests, 243 reviewer tests, 448 release-contract tests, full suite 2426 passed, and actionlint 1.7.12 clean.

**Comment by jhw7500** [comment]
@codex review

File: .github/workflows/gemini-auto-review.yml [hunk 5]
```diff
@@ -845,39 +1083,117 @@ jobs:
...
+                `- Run: ${trustedRun}`,
+                state.successful_head === null ? null : `- Reviewed: ${state.successful_head}`,
+                state.accepted_count === null ? null
+                  : `- Validation: accepted=${state.accepted_count}; filtered=${state.filtered_count}; normalized=${state.normalized_count}; filtered_max=${state.filtered_max_severity}`,
+                state.attempt_status === 'failure' ? `- Last attempt: failure (${trustedRun})` : null,
+              ].filter(Boolean).join('\n');
+            };
             const parseRecord = (comment) => {
-              if (comment.user?.type !== 'Bot') return null;
+              if (comment.user?.type !== 'Bot' || comment.user?.login !== botLogin) return null;
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/workflows/gemini-auto-review.yml]: Fixed in 12ef1f0e07066ff89af405676301ae261873f1a2. Gemini now preserves the existing v3 sticky across github_token and GitHub App publisher changes only when performed_via_github_app ID, slug, and bot login all match the explicitly bounded migration identity. Canonical callers retain publisher_app_id without retaining App credentials in token mode, and unrelated Bots or Apps remain rejected. Validation: full suite 2443 passed plus 48 subtests; release verifier 448 passed; actionlint 1.7.12 passed; pinned v1.46 commit-content verification passed.

**Comment by jhw7500** [comment]
@codex review

File: .github/actions/canonicalize-review/canonicalize_review.py [hunk 6]
```diff
@@ -0,0 +1,962 @@
...
+
+def _candidate_json_line(value: dict[str, object]) -> str:
+    """Serialize the exact literal JSON form accepted from a current candidate."""
+    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
+
+
+def _json_line(value: dict[str, object]) -> str:
+    return _candidate_json_line(value).translate(str.maketrans({
+        "<": "\\u003c", ">": "\\u003e", "&": "\\u0026",
+    }))
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/actions/canonicalize-review/canonicalize_review.py:377]: Fixed in 6d846e12bef26259d6b0c3254183c9a079ae890b. The canonical JSON renderer now escapes every Python splitlines boundary that JSON itself leaves literal: U+0085, U+2028, and U+2029. The regression matrix covers all ten splitlines boundaries and proves each rendered field remains one Markdown line while json.loads restores the exact path and quote. Validation: 10 focused boundary cases, 134 canonicalizer/scope tests, 448 release-verifier tests, full suite 2453 passed plus 48 subtests, actionlint 1.7.12, and pinned v1.46 commit-content verification all passed.

**Comment by jhw7500** [comment]
@codex review

File: .github/actions/canonicalize-review/canonicalize_review.py [hunk 7]
```diff
@@ -0,0 +1,963 @@
...
+
+def _candidate_json_line(value: dict[str, object]) -> str:
+    """Serialize the exact literal JSON form accepted from a current candidate."""
+    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
+
+
+def _json_line(value: dict[str, object]) -> str:
+    return _candidate_json_line(value).translate(str.maketrans({
+        "<": "\\u003c", ">": "\\u003e", "&": "\\u0026",
+        "\x85": "\\u0085", "\u2028": "\\u2028", "\u2029": "\\u2029",
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/actions/canonicalize-review/canonicalize_review.py:376]: Fixed in fcd60700d8ff5a426a749cad1f69117265bd477e. The canonicalizer now enforces a 64,000-byte post-render UTF-8 ceiling after JSON/HTML-safe escaping and returns hard candidate_oversize without publishing a canonical file. The existing publisher still independently preflights the complete 65,536-byte sticky envelope. Added an integrated < expansion regression, the existing near-60KB boundary assertion, and Claude/Gemini envelope coverage. Verification: 2456 tests + 48 subtests passed, canonicalizer/scope 135 passed, actionlint clean, and v1.46 commit-only verification passed.

**Comment by jhw7500** [comment]
@codex review

File: .github/actions/canonicalize-review/review_scope.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        if status not in set(STATUS_NAMES.values()) or not _safe_path(filename):
+            raise ScopeValidationError("invalid scope manifest file")
+        if status in {"renamed", "copied"}:
+            if not _safe_path(previous):
+                raise ScopeValidationError("invalid scope manifest rename")
+        elif previous is not None or "previous_filename" in file:
+            raise ScopeValidationError("invalid scope manifest file")
+        parsed_files.append(ScopeFile(status, filename, previous))
+    if not parsed_files or len({file.filename for file in parsed_files}) != len(parsed_files):
+        raise ScopeValidationError("invalid scope manifest file identity")
```
[the remark this answers is not in the document]
Comment (jhw7500) [reply .github/actions/canonicalize-review/review_scope.py]: Fixed in 8a2c0e2257a0b6699fb17896989cab5d482f55fa. The scope boundary now accepts the zero-record case only when all three trusted facts agree: diff_mode is full, the selected diff is exactly zero bytes, and immutable merge-base..head name-status reconstruction is empty. Empty delta, non-empty selected bytes, and an empty manifest hiding real Git records remain scope_invalid. The accepted scope exposes no changed anchors, so no actionable finding can be accepted. Added preparation, boundary, adversarial mismatch, and end-to-end clean canonicalization regressions. Verification: 163 focused tests, 2462 full-suite tests + 48 subtests, actionlint clean, and v1.46 commit-only verification passed.

**Comment by jhw7500** [comment]
@codex review

**Comment by jhw7500** [comment]
OpenCode run 32881946776 attempt 2 was manually adjudicated after the retry cap.

- Rejection: `anchor_out_of_scope` on `.github/actions/canonicalize-review/canonicalize_review.py:682`.
- Claim: candidate/previous-review paths could be redirected to arbitrary runner files.
- Adjudication: false positive. Both paths are fixed by trusted workflow code, not model/PR input; checkout-seeded output symlinks are removed immediately before generation; candidate reads use `O_NOFOLLOW` and require a bounded regular file. The previous review intentionally lives under `${RUNNER_TEMP}`, so constraining it to the repository root would break the valid contract. Strict candidate/scope parsing and exact previous-renderer validation also prevent raw arbitrary bytes from reaching publication.
- Gate behavior: the changed-line scope check correctly rejected the unsupported HIGH finding and preserved the prior successful sticky review.

Current HEAD `8a2c0e2257a0b6699fb17896989cab5d482f55fa`: full Test Fleet green, Gemini clean, Codex clean, all review threads resolved.
